feat: cache asm secrets

matt2e opened the pull request:
closes #1782
ASM changes:
- `ASM` uses `leader` package to coordinate between leader and followers
- The leader syncs secrets from ASM
    - Only requests secrets that have been updated since last sync
- Followers always uses admin service to get the secrets from the leader
- Both leader and follower cache secrets using `secretsCache`

Other changes:
- Remove additional db connection pools created at start up. Now a single db pool is created along with a dal at start up, which is then passed into the controller service and anything else that needs it

alecthomas commented:
Wasn't this going to be two separate PRs?

alecthomas commented:
Can you split it up please?

matt2e commented:
I'll add some more tests to this PR too

gak commented on common/configuration/asm_follower.go:
```diff
@@ -0,0 +1,97 @@
+	}
+	entries := []Entry{}
+	for _, s := range resp.Msg.Secrets {
+		components := strings.Split(s.RefPath, ".")
```
I've seen this pattern a few times, maybe a `fn (rp *) ToRef (Ref, error) {}` ?

matt2e replied:
I'll make a helper func for this 👍

matt2e replied:
`ParseRef` already exists 🤦‍♂️

matt2e commented:
> I'll add some more tests to this PR too

added integration test for syncing secrets

matt2e commented on common/configuration/asm_leader.go:
```diff
@@ -0,0 +1,315 @@
+	client *secretsmanager.Client
+
+	// indicates if the initial sync with ASM has finished
+	loaded atomic.Value[bool]
```
Oh nice

matt2e commented on common/configuration/asm_leader.go:
```diff
@@ -0,0 +1,315 @@
+
+			err := l.sync(ctx)
+			if err == nil {
+				backOff = syncInitialBackoff
```
This case is for when there is no error and is resetting the backoff. The error is logged below

matt2e commented:
Updated: 
- added `secretsCache` which `asmLeader` and `asmFollower` use
- made the integration test where set secrets via ftl and ASM directly handle both cases:
    - setting/getting via leader which calls ASM
    - setting/getting via a follower, which calls a leader, which calls ASM
